Refactor TestSource

jgraham opened the pull request:
Previously the TestSource subclasses were used for two things:
* As class objects, for getting information about test groups, and enqueueing test groups containing tests and their metadata.
* As class instances, for getting the tests out of the queue.

This was quite akward. In particular, all the functionality on the subclasses was actually related to the first use, and we ended up passing around tuple of (class, data) to all the places we needed to get information about tests groups.

Instead of this, rename TestSource to TestQueueBuilder and make all the class methods for the first use case into normal instance methods. That allows us to just pass around a class instance like normal.

For the second use case, create a new TestSource class inside testrunner which only knows how to get test groups out of the queue when needed.

As part of this refactoring we make a couple of other changes to improve clarity:

* Since we always enqueue all test groups before we try to read any, wrap the queue classes in wrappers that only give write and read access. This makes it clear that no writing will happen after the queue has been initialised.

* Rearrange some of the code in the TestRunnerManager initialization, so that we don't initalize internal objects (e.g. the TestSource) before we have a logger available.

jgraham committed:
Refactor TestSource

Previously the TestSource subclasses were used for two things:
* As class objects, for getting information about test groups, and
enqueueing test groups containing tests and their metadata.
* As class instances, for getting the tests out of the queue.

This was quite akward. In particular, all the functionality on the
subclasses was actually related to the first use, and we ended up
passing around tuple of (class, data) to all the places we needed to
get information about tests groups.

Instead of this, rename TestSource to TestQueueBuilder and make all
the class methods for the first use case into normal instance
methods. That allows us to just pass around a class instance like
normal.

For the second use case, create a new TestSource class inside
testrunner which only knows how to get test groups out of the queue
when needed.

As part of this refactoring we make a couple of other changes to
improve clarity:

* Since we always enqueue all test groups before we try to read any,
wrap the queue classes in wrappers that only give write and read
access. This makes it clear that no writing will happen after the
queue has been initialised.

* Rearrange some of the code in the TestRunnerManager initialization,
so that we don't initalize internal objects (e.g. the TestSource)
before we have a logger available.

WeizhongX commented on tools/wptrunner/wptrunner/testloader.py:
```diff
@@ -23,6 +24,8 @@
 manifest_update = None
 download_from_github = None
 
+TestsByType = Mapping[Tuple[str, str], List[wpttest.Test]]
```
This is tests by subsuite and type, right? Not sure if calling it `TestsBySubsuiteType` is too awkward. Maybe we can add a comment to make it clear what it is?

jonathan-j-lee commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -253,6 +253,24 @@ def is_alive(self):
+                self.logger.debug(f"Got new test group subsuite:{self.current_group[1]} "
+                                  f"test_type:{self.current_group[2]}")
+            except Empty:
+                self.logger.warning("Failed to get group from queue")
```
Should we still have this warning if an empty queue no longer represents an anomaly? This was removed in https://github.com/web-platform-tests/wpt/pull/42350#discussion_r1350547682 for being too noisy.

jgraham replied:
I deleted that and didn't save the file. Sigh :)

jonathan-j-lee reviewed (approved):
LGTM with nits, and the new type hints are appreciated.

jgraham committed:
Make TestRoot into a class

This is easier to type annotate than a TypedDict, and overall slightly
nicer design.

jgraham commented on tools/wptrunner/wptrunner/wptrunner.py:
```diff
@@ -47,7 +49,9 @@ def setup_logging(*args, **kwargs):
 
 
-def get_loader(test_paths, product, **kwargs):
+def get_loader(test_paths: Mapping[str, List[str]],
```
So, it turns out that Python 3.7 doesn't have `TypedDict`, so I've gone down the rabbit hole of trying to make this into an actual class instead. That's all in a new commit, so @jonathan-j-lee if you don't mind reviewing this again, it would be appreciated (hopefully I found all the corner cases now :) ).

jonathan-j-lee replied:
> So, it turns out that Python 3.7 doesn't have TypedDict

Ah, that's too bad; I would have been fine with an `Any` placeholder, had I known.

`TestRoot` class LGTM, thanks!
